src/Microsoft.AspNetCore.SignalR.Redis/Internal/RedisProtocol.cs

```diff
@@ -35,136 +39,172 @@ public RedisProtocol(IReadOnlyList<IHubProtocol> protocols)
+            IReadOnlyList<string> excludedIds = null;
+            var idCount = MsgPackUtil.ReadArrayHeader(ref data);
+            if (idCount > 0)
             {
-                IReadOnlyList<string> excludedIds = null;
-
-                var idCount = reader.ReadVarInt();
-                if (idCount > 0)
+                var ids = new string[idCount];
+                for (var i = 0; i < idCount; i++)
                 {
-                    var ids = new string[idCount];
-                    for (var i = 0; i < idCount; i++)
-                    {
-                        ids[i] = reader.ReadString();
-                    }
-
-                    excludedIds = ids;
+                    ids[i] = MsgPackUtil.ReadString(ref data);
                 }
 
-                var message = SerializedHubMessage.ReadAllSerializedVersions(reader);
-                return new RedisInvocation(message, excludedIds);
+                excludedIds = ids;
             }
+
+            // Read payload
+            var message = ReadSerializedHubMessage(ref data);
+            return new RedisInvocation(message, excludedIds);
         }
 
-        public RedisGroupCommand ReadGroupCommand(byte[] data)
+        public RedisGroupCommand ReadGroupCommand(ReadOnlyMemory<byte> data)
         {
             // See WriteGroupCommand for format.
-            using (var stream = new MemoryStream(data))
-            using (var reader = new BinaryReaderWithVarInt(stream, _utf8NoBom))
-            {
-                var id = reader.ReadVarInt();
-                var serverName = reader.ReadString();
-                var action = (GroupAction)reader.ReadByte();
-                var groupName = reader.ReadString();
-                var connectionId = reader.ReadString();
+            ValidateArraySize(ref data, 5, "GroupCommand");
 
-                return new RedisGroupCommand(id, serverName, action, groupName, connectionId);
-            }
+            var id = MsgPackUtil.ReadInt32(ref data);
+            var serverName = MsgPackUtil.ReadString(ref data);
+            var action = (GroupAction)MsgPackUtil.ReadByte(ref data);
+            var groupName = MsgPackUtil.ReadString(ref data);
+            var connectionId = MsgPackUtil.ReadString(ref data);
+
+            return new RedisGroupCommand(id, serverName, action, groupName, connectionId);
         }
 
-        public int ReadAck(byte[] data)
+        public int ReadAck(ReadOnlyMemory<byte> data)
         {
             // See WriteAck for format
-            using (var stream = new MemoryStream(data))
-            using (var reader = new BinaryReaderWithVarInt(stream, _utf8NoBom))
+            ValidateArraySize(ref data, 1, "Ack");
+            return MsgPackUtil.ReadInt32(ref data);
+        }
+
+        private void WriteSerializedHubMessage(Stream stream, SerializedHubMessage message)
+        {
+            // Written as a MessagePack 'map' where the keys are the name of the protocol (as a MessagePack 'str')
```

Comment:
Why not an array of pairs?

Reply:
Identical size and this is more "accurate"

Reply:
A map is literally encoded as an array of pairs but with a different header value

Reply:
I see